File: 623-add-one-row-to-tree/623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(
        self, root: Optional[TreeNode], val: int, depth: int, left=True
    ) -> Optional[TreeNode]:
        if depth == 1:  
            return TreeNode(val=val, left=root) if left else TreeNode(val=val, right=root)
        
        if not root:
            return None
          
        root.left = self.addOneRow(root.left, val, depth - 1, left=True)
        root.right = self.addOneRow(root.right, val, depth - 1, left=False)

        return root
        
        
        
            
        
        

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun addOneRow(
        node: TreeNode?, value: Int, depth: Int, isleft: Boolean = true
    ): TreeNode? {
        if (depth == 1) return TreeNode(value).apply {
            if (isleft) left = node else right = node
        }
        node ?: return null
        return node.apply {
            left = addOneRow(node.left, value, depth - 1, true)
            right = addOneRow(node.right, value, depth - 1, false)
        }
    }
}